components/glhelper.py

```diff
@@ -0,0 +1,244 @@
+#!/usr/bin/env python3
+"""Helper functions for quaternion math, arcball rotation, and paths.
+TODO: Give attribution to Printrun
+"""
+
+import math
+import functools
+import numpy as np
+from OpenGL.GL import (GL_VERTEX_ARRAY, GL_FLOAT, GL_LINE_STRIP,
+                       glEnableClientState, glVertexPointer,
+                       glDrawArrays, glDisableClientState)
+
+
+def arcball(p1x, p1y, p2x, p2y, r):
+    """Update arcball rotation."""
+    last = sphere_coords(p1x, p1y, r)
+    cur = sphere_coords(p2x, p2y, r)
+    rot_axis = np.cross(cur, last)
+
+    # calculate angle between last and cur
+    d = map(lambda x, y: x - y, last, cur)
+    t = math.sqrt(sum(x*x for x in d)) / (2.0 * r)
+    if t > 1.0:
+        t = 1.0
+    if t < -1.0:
+        t = -1.0
+    phi = 2.0 * math.asin(t)
+
+    return vector_to_quat(rot_axis, phi)
+
+
+def sphere_coords(x, y, r):
+    """Compute the intersection from screen coords to the arcball sphere."""
+    d2 = x*x + y*y
+    r2 = r * r
+    if math.sqrt(d2) <= r * 0.70710678118654752440:

```

Comment:
What does 0.70710678118654752440 and r * 0.70710678118654752440 represents?

Reply:
This achieves what is described [here](https://www.khronos.org/opengl/wiki/Object_Mouse_Trackball#Of_mice_and_manipulation). To smooth the transition from dragging in the sphere vs out, we define a piecewise function between the sphere and the hyperbolic sheet, where sqrt(x^2+y^2) <= sqrt(r^2/2), and sqrt(r^2/2) = r * 0.707106...